File: frontend/app/api/transaction.ts
import apiFetch from "./api";

export type AcctTransaction = {
  ACCT_TRANS_NUM: number;
  ACCT_TRANS_DATE: Date | null;
  CUST_NUMBER: number | null;
  ACCT_TRANS_TYPE: string | null;
  ACCT_TRANS_AMOUNT: number | null;
};

export type AcctTransactionUpdate = {
  ACCT_TRANS_DATE: Date;
  CUST_NUMBER: number;
  ACCT_TRANS_TYPE: string;
  ACCT_TRANS_AMOUNT: number;
};

export async function findAllAcctTransactions() {
  const response = await apiFetch("/transactions/all");

  if (!response) {
    throw new Error("Failed to retrieve transactions.");
  }

  const data = (await response.json()) as AcctTransaction[];
  return data;
}

export async function saveTransaction(newTransaction: AcctTransaction) {
  const response = await apiFetch("/transactions/save", {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify(newTransaction),
  });

  if (!response) {
    throw new Error("Failed to save transaction.");
  }

  return response.ok;
}

export async function updateTransaction(transactionId: number, updatedTransaction: AcctTransactionUpdate) {
  const response = await apiFetch(`/transactions/update/${transactionId}`, {
    method: "PUT",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify(updatedTransaction),
  });

  if (!response) {
    throw new Error("Failed to update transaction.");
  }

  return response.ok;
}

export async function deleteTransactionById(transactionId: number) {
  const response = await apiFetch(`/transactions/delete/${transactionId}`, {
    method: "DELETE",
  });

  if (!response) {
    throw new Error("Failed to delete transaction.");
  }

  return response.ok;
}

File: backend/app/api/routes/transaction.py
from fastapi import APIRouter, Depends, HTTPException

from app.database import get_db
from app.models import TransactionCreate, TransactionUpdate

router = APIRouter(prefix="/transactions")


@router.get("/all")
def get_all_transactions(db=Depends(get_db)):
    cursor = db.cursor(dictionary=True)

    query = "SELECT * FROM ACCT_TRANSACTION"
    cursor.execute(query)
    return cursor.fetchall()


@router.post("/save", status_code=201)
def save_transaction(transaction: TransactionCreate, db=Depends(get_db)):
    transaction_data = (
        transaction.ACCT_TRANS_NUM,
        transaction.ACCT_TRANS_DATE,
        transaction.CUST_NUMBER,
        transaction.ACCT_TRANS_TYPE,
        transaction.ACCT_TRANS_AMOUNT,
    )

    try:
        cursor = db.cursor()
        cursor.execute(
            """
            INSERT INTO ACCT_TRANSACTION
            (ACCT_TRANS_NUM, ACCT_TRANS_DATE, CUST_NUMBER, ACCT_TRANS_TYPE, ACCT_TRANS_AMOUNT)
            VALUES (%s, %s, %s, %s, %s)
            """,
            transaction_data,
        )

        db.commit()
    except:
        HTTPException(status_code=500, detail="Failed to save transaction")


@router.put("/update/{id}")
def update_transaction(id: int, transaction: TransactionUpdate, db=Depends(get_db)):
    transaction_data = (
        transaction.ACCT_TRANS_DATE,
        transaction.CUST_NUMBER,
        transaction.ACCT_TRANS_TYPE,
        transaction.ACCT_TRANS_AMOUNT,
    )
    try:
        cursor = db.cursor(dictionary=True)
        cursor.execute(
            """
            UPDATE ACCT_TRANSACTION
            SET ACCT_TRANS_DATE = %s, CUST_NUMBER = %s, ACCT_TRANS_TYPE = %s, ACCT_TRANS_AMOUNT = %s
            WHERE ACCT_TRANS_NUM = %s
            """,
            (*transaction_data, id),
        )
        db.commit()
    except:
        raise HTTPException(status_code=500, detail="Failed to update transaction")


@router.delete("/delete/{transaction_id}", status_code=204)
def delete_transaction_by_id(transaction_id: int, db=Depends(get_db)):
    cursor = db.cursor()

    cursor.execute(
        """
        DELETE FROM ACCT_TRANSACTION WHERE ACCT_TRANS_NUM = %s""",
        [transaction_id],
    )

    db.commit()
